java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java

```diff
@@ -356,5 +399,17 @@ void updateChangedSources(Set<File> changedFiles) {
         state = State.TOO_MANY_CHANGES;
       }
     }
+
+    void checkForNonCompilableChanges(Set<File> changedNonCompileFiles) {
+      if (state.equals(State.WAITING_FOR_SOURCES)) {
```

Comment:
Why only in WAITING_FOR_SOURCES? What about Collecting?

Reply:
As mentioned in the other comment, it's for the case, were proto file changes were made, but the state is WAITING_FOR_SOURCES, so they cannot be compared to actual sources, and if getAndResetChangedSources is called before the sources are collected, full compile will have to happen. If the sources were collected, the files in changedNonCompilableSources property will be checked against them, and the state will be set to NON_COMPILABLE_CHANGES. As well as if the state was COLLECTING, and any proto files from the sources were changed, the NON_COMPILABLE_CHANGES would be set too.

Reply:
I think this isn't a problem. This can happen also for scala and java files, right? That they're called when WAITING_FOR_SOURCES is the state. What's the problem?

Reply:
It is not a problem. It's just the part, that makes the implementation less straight forward, as you can see, the WAITING_FOR_SOURCES state jumps out in multiple methods.